### ISSUE TITLE
Curve_fit act differently after shuffling data.

### ISSUE THREAD

**Issue by waystogetthere** [scipy/scipy#17926]
Hello,

I have a weird problem with the function '_curve_fit_'.

It is a regression problem, given the feature matrix X with shape: [833, 3], where 833 is the number of samples, and 3 is the dimension of features, we want to predict real-value y: (833,)

The X has 3 columns: [a1, a2, a3], the fitted function is:

`y = a3 *  np.exp(k0 * a2 * (k1 * a1 + b)),`

where k0, k1, and b are the learnable parameters.

THE PROBLEM IS:

when first time fitting the model it always comes to a result that all predicted $\hat{y}$ are close to 0.
when I shuffle the training set (both X and y), the problem solves.

**Comment by waystogetthere** [comment]
The data is here , along with my code.
https://www.dropbox.com/scl/fo/n3tm6k78yk0je1jc8spfy/h?dl=0&rlkey=bqkgf4k9jj6xj7m5p4w6w7jao

**Comment by waystogetthere** [comment]
```
from scipy.optimize import curve_fit
import numpy as np
from sklearn.metrics import mean_absolute_error



def func_exp_linear(X, k0, k1, b):
    T = X[:, 0]  # Temp
    t = X[:, 1]  # time
    res_cl = X[:, 2]  # res cl
    return res_cl * np.exp(k0 * t * (k1 * T + b))

with open('X.npy', 'rb') as f_x:
    X = np.load(f_x)
    # X shape: (Num_samples, 3), 3 dimensions: T, t, res_cl, whereas T has been preprocessed(standerized).
with open('y.npy', 'rb') as f_y:
    y = np.load(f_y)
    # y shape; (Num_samples, )

######
popt, pcov = curve_fit(func_exp_linear, X, y)
y_hat = func_exp_linear(X, *popt)
MAE = mean_absolute_error(y, y_hat)


while MAE > 0.8:  # start shuffle
    print('the training MAE is too large!', MAE, 'start shuffle')
    n_obs = len(y)

    all_ = np.concatenate((X, y.reshape(-1, 1)), axis=1)  # concat X and y
    index = np.arange(n_obs).reshape(-1, 1)  # index indicates original order
    all_ = np.concatenate((index, all_), axis=1)  # index is at the first column

    np.random.shuffle(all_)
    X_new = all_[:, 1:4]  # T, t, res_cl
    y_new = all_[:, -1]  # tap_cl

    popt, pcov = curve_fit(func_exp_linear, X_new, y_new)
    y_hat_new = func_exp_linear(X_new, *popt)
    MAE = mean_absolute_error(y_hat_new, y_new)

    print('After shuffling the new MAE is: ', MAE)
    for i in range(n_obs): # index is used to double-check that shuffled data have the same content with orignal data
        ind = int(all_[i, 0])
        assert (X_new[i, :] == X[int(ind), :]).all()
        assert (y_new[i] == y[int(ind)]).all()


import matplotlib.pyplot as plt
plt.figure()
plt.plot(y_hat, label='yhat')
plt.plot(y_hat_new, label='yhat new')
plt.plot(y, label='y')
plt.legend()
plt.savefig('Predict Result.png', dpi=300)

```

**Comment by waystogetthere** [comment]
This is the result, 'y' is the ground truth. 'yhat' is the first fitted result. 'yhat_new' is the result after the data is shuffled(maybe more than 1 time) 
<img width="1062" alt="image" src="https://user-images.githubusercontent.com/30395015/216750477-894b7d46-bdf4-44cd-b9bf-fbcf2e4fe88e.png">

**Comment by mdhaber** [comment]
It looks like there are only two independent parameters. Please try eliminating one.

**Comment by waystogetthere** [comment]
> It looks like there are only two independent parameters. Please try eliminating one and seeing if this fixes the problem.

Hi Sir, thanks for your fast reply! Do you mean that one of the parameters is redundant: k0, k1, b?

**Comment by rkern** [comment]
Yes, expand `k0 * a2 * (k1 * a1 + b)` and you will see that you only have two independent parameters: `c0 = k0*k1` and `c1 = k0*b`.

**Comment by waystogetthere** [comment]
YESS! After eliminating k0 the problem solves, now when it first train the model it comes to a reasonable result.

So will the redundant parameter severely affect the performance of  'curve_fit', and cause fatal outcome?

Thanks a lot. Never thought about this before

**Comment by mdhaber** [comment]
Great. Yes, the problem becomes challenging numerically when there is a redundant parameter.
